File: SC2001/Lab/Project 2/dijkstra_adjmatrix.c
#include <stdio.h>
#include <stdlib.h>

#define VER 10

int curSize = 0;

void printArray(int *array, int n, int m)
{
    for (int i = n; i <= m; i++)
    {
        printf("%d ", array[i]);
    }
    printf("\n");
}

struct Graph
{
    int V;
    int **array;
};

struct Graph *createGraph(int V)
{
    struct Graph *graph = (struct Graph *)
        malloc(sizeof(struct Graph));
    graph->V = V;
    graph->array = (int **)malloc(V * sizeof(int *));
    for (int i = 0; i < V; i++)
    {
        graph->array[i] = (int *)malloc(V * sizeof(int));
        for (int j = 0; j < V; j++)
        {
            graph->array[i][j] = -1;
        }
    }
    return graph;
}

void printGraphMatrix(struct Graph *graph, int vertices)
{
    for (int i = 0; i < vertices; i++)
    {
        for (int j = 0; j < vertices; j++)
        {
            printf("%d ", graph->array[i][j]);
        }
        printf("\n");
    }
}

void addEdge(struct Graph *graph, int src, int dest, int weight)
{
    graph->array[src][dest] = weight;
    graph->array[dest][src] = weight;
}

// Priority Queue

struct Node
{
    int dest;
    int weight;
};

void insert(struct Node *PQ, int dest, int totalWeight)
{
    struct Node new;
    new.dest = dest;
    new.weight = totalWeight;
    PQ[curSize] = new;
    curSize++;
}

struct Node extractMin(struct Node *PQ)
{
    int min = PQ[0].weight;
    int minDex = 0;
    for (int i = 0; i < curSize; i++)
    {
        if (PQ[i].weight < min)
        {
            min = PQ[i].weight;
            minDex = i;
        }
    }
    struct Node returnNode = PQ[minDex];
    for (int i = minDex; i < curSize; i++)
    {
        PQ[i] = PQ[i + 1];
    }
    curSize--;
    return returnNode;
}

void dijkstra(struct Graph *g, int *dist)
{
    struct Node PQ[500];
    int visited[VER];
    for (int i = 0; i < VER; i++)
    {
        visited[i] = 0;
        dist[i] = 99999999;
    }
    dist[0] = 0;
    insert(PQ, 0, 0);
    while (curSize > 0)
    {
        struct Node currentNode = extractMin(PQ);
        visited[currentNode.dest] = 1;
        for (int i = 0; i < VER; i++)
        {

            if (g->array[currentNode.dest][i] != -1)
            {
                if (visited[i] != 1)
                {
                    int newDist = dist[currentNode.dest] + g->array[currentNode.dest][i];
                    if (newDist < dist[i])
                    {
                        dist[i] = newDist;
                        insert(PQ, i, newDist);
                    }
                }
            }
        }
    }
    printArray(visited, 0, VER - 1);
}

int main()
{
    struct Graph *g = createGraph(VER);
    addEdge(g, 0, 1, 4);
    addEdge(g, 0, 7, 8);
    addEdge(g, 1, 2, 8);
    addEdge(g, 1, 7, 11);
    addEdge(g, 2, 3, 7);
    addEdge(g, 2, 8, 2);
    addEdge(g, 2, 5, 4);
    addEdge(g, 3, 4, 9);
    addEdge(g, 3, 5, 14);
    addEdge(g, 4, 5, 10);
    addEdge(g, 5, 6, 2);
    addEdge(g, 6, 8, 6);
    addEdge(g, 6, 7, 1);
    addEdge(g, 7, 8, 7);
    int dist[VER];
    dijkstra(g, dist);
    printGraphMatrix(g, VER);
    printArray(dist, 0, VER - 1);
}

File: SC2001/Lab/Project 2/dijkstra_adjmatrix.py
class PriorityQueue:
    def __init__(self):
        self.queue = []
        self.length = 0

    def insert(self, dest, totalWeight):
        self.queue.append({'dest': dest, 'weight': totalWeight})
        self.length += 1

    def extractMin(self):
        minimum = self.queue[0]['weight']
        mIndex = 0
        for i in range(self.length):
            if self.queue[i]['weight'] < minimum:
                minimum = self.queue[i]['weight']
                mIndex = i
        returnNode = self.queue[mIndex]
        for i in range(mIndex, self.length - 1):
            self.queue[i] = self.queue[i + 1]
        self.queue.pop()
        self.length -= 1
        return returnNode


class Graph:
    def __init__(self, numVertices):
        self.aMatrix = []
        self.dist = []
        self.visited = []
        self.numVertices = numVertices
        self.PQ = PriorityQueue()
        for i in range(numVertices):
            self.aMatrix.append([])
            self.visited.append(0)
            self.dist.append(999999999)
            for _ in range(numVertices):
                self.aMatrix[i].append(-1)

    # Unidirectional
    def addEdge(self, fr, to, weight):
        self.aMatrix[fr][to] = weight

    def printGraphMatrix(self):
        for i in range(self.numVertices):
            print(self.aMatrix[i])

    def dijkstra(self):
        self.dist[0] = 0
        self.PQ.insert(0, 0)
        while(self.PQ.length > 0):
            pqMin = self.PQ.extractMin()
            index = pqMin['dest']
            self.visited[index] = 1
            for vertice in range(self.numVertices):
                if self.aMatrix[index][vertice] != -1:
                    if self.visited[vertice] != 1:
                        newDist = self.dist[index] + \
                            self.aMatrix[index][vertice]
                        if newDist < self.dist[vertice]:
                            self.dist[vertice] = newDist
                            self.PQ.insert(vertice, newDist)
        print(self.dist)


if __name__ == '__main__':
    g = Graph(10)
    g.addEdge(0, 1, 4)
    g.addEdge(0, 7, 8)
    g.addEdge(1, 2, 8)
    g.addEdge(1, 7, 11)
    g.addEdge(2, 3, 7)
    g.addEdge(2, 8, 2)
    g.addEdge(2, 5, 4)
    g.addEdge(3, 4, 9)
    g.addEdge(3, 5, 14)
    g.addEdge(4, 5, 10)
    g.addEdge(5, 6, 2)
    g.addEdge(6, 8, 6)
    g.addEdge(6, 7, 1)
    g.addEdge(7, 8, 7)
    g.printGraphMatrix()
    g.dijkstra()
